### ISSUE TITLE
Setting random seeds

### ISSUE THREAD

**Issue by sebo361** [DeepLabCut/DeepLabCut#264]
Hi guys,

i am wondering if there is a possibility of making the DLC model deterministic by setting random seeds? So within the same settings, the model should give the same results e.g. loss value at 1st iteration. 
I added ``tf.set_random_seed(1)`` in different positions of ``train_network()`` but it didnt work so far..

Thanks! Sebastian

**Comment by AlexEMG** [comment]
That's not implemented currently.

**Comment by sebo361** [comment]
Can you maybe give me hints on how to implement it in DLC? I tried to use the `TF_CUDNN_DETERMINISTIC` flag as used [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/cudnn_determinism_test.py) but I could not get deterministic behavior. Which part exactly brings stochasticity in within DLC?

**Comment by AlexEMG** [comment]
You mean for training? You can set shuffle=False (in config file), then the loader loads the data in a particular sequence (I think based on index, but you need to check). In general, in the loader, i.e. here: dataset/pose_dataset.py you can change/freeze the training schedule..

https://github.com/AlexEMG/DeepLabCut/blob/887814e960ef70d193745c9716c1cf9a871abe33/deeplabcut/pose_estimation_tensorflow/dataset/pose_dataset.py#L76

**Comment by sebo361** [comment]
Yes exactly for training. Cool thanks for the hint - I used `np.random.seed(0)` in [line 140 of dataset/pose_dataset.py](https://github.com/AlexEMG/DeepLabCut/blob/20bb84dcdb740a73c8644d8abe70c5b20e64078f/deeplabcut/pose_estimation_tensorflow/dataset/pose_dataset.py#L140) which is probably the same as setting shuffle=False in config file. I still don't get deterministic behavior as the loss value of the first iteration varies when I start training the model. When using the same pretrained weights and same input images, should the loss not be the same?

The aim of making the code deterministic is that I can compare better different settings for training as the training results are 1:1 reproducible.

**Comment by sebo361** [comment]
Hi @AlexEMG, 

I am using `shuffle=False` and i am training on CPU only to avoid stochasticity of CUDNN when training on GPU. For training, I use the following config setup:

```
Config:
{'all_joints': [[0]],
 'all_joints_names': ['Hand'],
 'batch_size': 1,
 'bottomheight': 400,
 'crop': True,
 'crop_pad': 0,
 'cropratio': 0.4,
 'dataset': 'training-datasets/iteration-0/UnaugmentedDataSet_test1May17/test1_seb20shuffle1.mat',
 'dataset_type': 'default',
 'display_iters': 2,
 'fg_fraction': 0.25,
 'global_scale': 0.8,
 'init_weights': '/Users/sebo/DeepLabCut/deeplabcut/pose_estimation_tensorflow/models/pretrained/resnet_v1_50.ckpt',
 'intermediate_supervision': False,
 'intermediate_supervision_layer': 12,
 'leftwidth': 400,
 'location_refinement': True,
 'locref_huber_loss': True,
 'locref_loss_weight': 0.05,
 'locref_stdev': 7.2801,
 'log_dir': 'log',
 'max_input_size': 1500,
 'mean_pixel': [123.68, 116.779, 103.939],
 'metadataset': 'training-datasets/iteration-0/UnaugmentedDataSet_test1May17/Documentation_data-test1_20shuffle1.pickle',
 'min_input_size': 64,
 'minsize': 100,
 'mirror': False,
 'multi_step': [[0.005, 10000],
                [0.02, 430000],
                [0.002, 730000],
                [0.001, 1030000]],
 'net_type': 'resnet_50',
 'num_joints': 1,
 'optimizer': 'sgd',
 'pos_dist_thresh': 17,
 'project_path': '/Users/sebo/DeepLabCut/test1-seb-2019-05-17',
 'regularize': False,
 'rightwidth': 400,
 'save_iters': 50000,
 'scale_jitter_lo': 0.5,
 'scale_jitter_up': 1.25,
 'scoremap_dir': 'test',
 'shuffle': False,
 'snapshot_prefix': '/Users/sebo/DeepLabCut/test1-seb-2019-05-17/dlc-models/iteration-0/test1May17-trainset20shuffle1/train/snapshot',
 'stride': 8.0,
 'topheight': 400,
 'use_gt_segm': False,
 'video': False,
 'video_batch': False,
 'weigh_negatives': False,
 'weigh_only_present_joints': False,
 'weigh_part_predictions': False,
 'weight_decay': 0.0001}
INFO:tensorflow:Restoring parameters from /Users/sebo/DeepLabCut/deeplabcut/pose_estimation_tensorflow/models/pretrained/resnet_v1_50.ckpt
Training parameter:
{'stride': 8.0, 'weigh_part_predictions': False, 'weigh_negatives': False, 'fg_fraction': 0.25, 'weigh_only_present_joints': False, 'mean_pixel': [123.68, 116.779, 103.939], 'shuffle': False, 'snapshot_prefix': '/Users/sebo/DeepLabCut/test1-seb-2019-05-17/dlc-models/iteration-0/test1May17-trainset20shuffle1/train/snapshot', 'log_dir': 'log', 'global_scale': 0.8, 'location_refinement': True, 'locref_stdev': 7.2801, 'locref_loss_weight': 0.05, 'locref_huber_loss': True, 'optimizer': 'sgd', 'intermediate_supervision': False, 'intermediate_supervision_layer': 12, 'regularize': False, 'weight_decay': 0.0001, 'mirror': False, 'crop_pad': 0, 'scoremap_dir': 'test', 'dataset_type': 'default', 'use_gt_segm': False, 'batch_size': 1, 'video': False, 'video_batch': False, 'crop': True, 'cropratio': 0.4, 'minsize': 100, 'leftwidth': 400, 'rightwidth': 400, 'topheight': 400, 'bottomheight': 400, 'all_joints': [[0]], 'all_joints_names': ['Hand'], 'dataset': 'training-datasets/iteration-0/UnaugmentedDataSet_test1May17/test1_seb20shuffle1.mat', 'display_iters': 2, 'init_weights': '/Users/sebo/DeepLabCut/deeplabcut/pose_estimation_tensorflow/models/pretrained/resnet_v1_50.ckpt', 'max_input_size': 1500, 'metadataset': 'training-datasets/iteration-0/UnaugmentedDataSet_test1May17/Documentation_data-test1_20shuffle1.pickle', 'min_input_size': 64, 'multi_step': [[0.005, 10000], [0.02, 430000], [0.002, 730000], [0.001, 1030000]], 'net_type': 'resnet_50', 'num_joints': 1, 'pos_dist_thresh': 17, 'project_path': '/Users/sebo/DeepLabCut/test1-seb-2019-05-17', 'save_iters': 50000, 'scale_jitter_lo': 0.5, 'scale_jitter_up': 1.25}
```

When i start training with identical setup on CPU, loss outputs don't match with each other, which means that the model does not behave deterministically. Do you have an idea which element of the code still brings in stochasticity? 

Thanks already!

**Comment by sebo361** [comment]
Hi @MMathisLab & @AlexEMG,

i finally managed to make the code deterministic on the CPU. When training on a GPU, cuDNN still brings in stochastic values, but the standard deviation of losses is really low as shown in my quick evaluation :)

Does it sound interesting to you if I implement the option "deterministic = True/False" in deeplabcut.train()? Or might the option better be in the pose_cfg file?

[CPUvsGPU_loss.pdf](https://github.com/AlexEMG/DeepLabCut/files/3291582/CPUvsGPU_loss.pdf)
